libs/contrib/Data/Fun/Extra.idr

```diff
@@ -0,0 +1,54 @@
+module Data.Fun.Extra
+
+import Data.Fun
+import Data.HVect
+
+||| Apply an n-ary function to an n-ary tuple of inputs
+public export
+apply : {ts : Vect n Type} -> {cod : Type} -> Fun ts cod -> HVect ts -> cod
```

Comment:
I would call this `curry` and also define `uncurry`. Note that both `ts` and
`cod` can probably be given a `0` annotation (we only need to be able to match on `n`).

```idris
uncurry : {ts : Vect n Type} -> {cod : Type} -> (HVect ts -> cod) -> Fun ts cod
```

Reply:
Done, though I would call them `uncurry` and `curry` instead (opposite of what you wrote).

Reply:
Re. annotation: I'm never sure what one should match on. 
Should vectors carry their length and be packed? 
Should vectors be linked lists and we should only calculate their length if we want to?
Should we be able to pass constraints like `{wn n : Nat} -> {wts ts : Vect n Type} -> {0 auto atLeastOne : NonZero [wn, wt]}` that allows the type-checker to match on the pattern:
```idris
{n = 0} {ts : []}
{n = S n} {ts : x :: xs}
```
because each invocation is guaranteed to have a case tree (but we don't know right now which one it should be)?

Reply:
A `Vect n Type` is essentially a linked list of erased stuff.
It's a lot cheaper to just carry the `n` around hence my suggestion.

Reply:
At some point someone needs to compute this `n`. Merely having a `Vect n` doesn't give you it.

Reply:
Yes. I am arguing in favour of carrying the `n` around instead
of carrying the `Vect n Type` around. One Integer vs. a linked
list of erased things.

Reply:
Not quite, if you're not careful you'll be carrying one integer for every cons-cell.

Reply:
Even more reasons to make it `{0 ts : Vect n Type}`.

Reply:
This is frustrating. This means there's something I'm not getting, perhaps an assumption I'm not even realising I'm making.

At runtime, is a `Nat` represented in binary or as a linked list of `Succ` cells?

Reply:
At runtime a `Nat` is an `Integer` i.e. it's stored in binary.

Reply:
OK, so that's not my implicit assumption.

So we have a vector, i.e., linked list, somewhere. The linked list is not carrying its own length.
If I require library functions to have the length as relevant, then we'd be forced to calculate the length of the vector, on top of having the vector already, every time we call the library function.

Reply:
I don't agree with the assumption that we necessarily are carrying around
the linked list. It does not contain anything of runtime interest (it's only
`Type`s!) except for its length which is implicit in the structure.

So we may as well pass the length around and let the vector be erased!

Reply:
The vector isn't going to be erased, it contains the data we care about (in other places in the code-base).
It's just that if we make it irrelevant, we won't pass around a pointer for its different cells, and we'll need to somehow get its length.
Which means either calculating it for a long vector, or amortizing the calculation by allocating a length for each cell.

Reply:
OK, I think I understand. Your point is that we don't need the types at runtime, just the number of arguments.

Reply:
Got it. Changed it now. 
Thanks for the patience, I appreciate it a lot.

Reply:
I really like that QTT forces us to think harder about all of this!